idaes/core/util/model_diagnostics.py

```diff
@@ -3762,52 +3764,64 @@ def check_parallel_jacobian(
-            # idx is the "local index", uidx is the "global index"
-            # Frobenius norm of the matrix is 2-norm of this column vector
-            unorm = norm(u, ord="fro")
-            for v, vidx in vecs[idx + 1 :]:
-                vnorm = norm(v, ord="fro")
-
-                # Explicitly multiply a row vector * column vector
-                prod = u.transpose().dot(v)
-                absprod = abs(prod[0, 0])
-                diff = abs(absprod - unorm * vnorm)
-                if diff <= tolerance or diff <= tolerance * max(unorm, vnorm):
-                    parallel.append((uidx, vidx))
-
-    parallel = [(components[uidx], components[vidx]) for uidx, vidx in parallel]
+        mat = jac.transpose().tocsr()
+
+    # Take product of all rows/columns with all rows/columns by taking outer
+    # product of matrix with itself
+    outer = mat @ mat.transpose()
+
+    # Along the diagonal of the outer product you get the dot product of the row
+    # with itself, which is equal to the norm squared.
+    norms = np.sqrt(outer.diagonal())
+
+    # Want to ignore indices with zero norm. By zeroing out the corresponding
+    # entries of the scaling matrix, we set the corresponding rows and columns
+    # to zero, which will then be ignored.
+
+    zero_norm_indices = np.nonzero(np.abs(norms) <= zero_norm_tolerance)
+    inv_norms = 1 / norms
+    inv_norms[zero_norm_indices] = 0
+
+    # Divide each row and each column by the vector norm. This leaves
+    # the entries as dot(u, v) / (norm(u) * norm(v)). The exception is
+    # entries with "zero norm", whose corresponding rows and columns are
+    # set to zero.
+    scaling = diags(inv_norms)
+    outer = scaling * outer * scaling
+
+    # Get rid of duplicate values by only taking upper triangular part of
+    # resulting matrix
+    upper_tri = triu(outer)
+
+    # Set diagonal elements to zero
+    # Subtracting diags(upper_tri.diagonal()) is a more reliable
+    # method of getting the entries to exactly zero than subtracting
+    # an identity matrix, where one can encounter values of 1e-16
+    upper_tri = upper_tri - diags(upper_tri.diagonal())
+
+    # Get the nonzero entries of upper_tri in three arrays,
+    # corresponding to row indices, column indices, and values
+    rows, columns, values = find(upper_tri)
+
+    # We have that dot(u,v) == norm(u) * norm(v) * cos(theta) in which
+    # theta is the angle between u and v. If theta is approximately
+    # 0 or pi, sqrt(2*(1 - abs(dot(u,v)) / (norm(u) * norm(v)))) approximately
+    # equals the number of radians from 0 or pi. A tolerance of 1e-8 corresponds
+    # to a tolerance of sqrt(2) * 1e-4 radians
+
+    # The expression (1 - abs(values) < tolerance) returns an array with values
+    # of ones and zeros, depending on whether the condition is fulfilled or not.
+    # We then find indices where it is filled using np.nonzero.
+    parallel_1D = np.nonzero(1 - abs(values) < tolerance)[0]
```

Comment:
Both the previous implementation and the Gurobi implementation checked against a semi-normalized epsilon:

```
                #
                # Normalize the tolerance in the test; want the same
                # result for vectors u and v and 1000u and 1000v
                #
                if abs(abs(dotprod) - norm1 * norm2) > partol * norm1:
                    continue
```
Note that this formulation doesn't scale the tolerance with norm2 for some reason. Also, possibly for reasons of efficiency, they use 1-norms instead of 2-norms. That's a false economy, though, because it's no longer guaranteed that `abs(dotprod) <= norm1 * norm2` (Cauchy-Schwartz inequality). The equation we're using comes from the formula for angle between two vectors: `u.dot(v) = unorm*vnorm*cos(theta)` along with Taylor expansions around `theta=0` and `theta=pi`.

We could implement the check like `unorm*vnorm - abs(u.dot(v)) < tolerance*unorm*vnorm`, but that has the disadvantage of being harder to vectorize. Floating point division, even division by some number on the order of 1e-16, is well defined and can be computed accurately. From Wikipedia:

> There are no cancellation or absorption problems with multiplication or division, though small errors may accumulate as operations are performed in succession.

It's *addition and subtraction* that can cause problems with catastrophic cancellation.

Reply:
I would advocate for `unorm*vnorm - abs(u.dot(v)) < tolerance*max(1, unorm*vnorm)`, where all norms are 2-norms. Point taken on division, but the above is still fewer operations than the current implementation.